File: 06 - LOGICIEL EMBARQUE/firmware-esp32/components/hardware_contracts/hardware_contracts.cpp
#include "hardware_contracts.hpp"

#include <algorithm>
#include <cmath>

namespace ventilation {

uint16_t dac_word_from_percent(float percent) {
  const float bounded = std::clamp(percent, 0.0F, 100.0F);
  const float millivolts = bounded * 100.0F;
  const auto code12 = static_cast<uint16_t>((millivolts / 10000.0F) * 4095.0F);
  return static_cast<uint16_t>(code12 << 4U);
}

DacFrame dac_frame_from_percent(float percent, uint8_t channel) {
  const uint16_t word = dac_word_from_percent(percent);
  const uint8_t low = static_cast<uint8_t>(word & 0xffU);
  const uint8_t high = static_cast<uint8_t>((word >> 8U) & 0xffU);
  if (channel == 0U) return {0x02U, {low, high, 0U, 0U}, 2U};
  if (channel == 1U) return {0x04U, {low, high, 0U, 0U}, 2U};
  if (channel == 2U) return {0x02U, {low, high, low, high}, 4U};
  return {0x00U, {0U, 0U, 0U, 0U}, 0U};
}

SensorSampleStatus classify_scd41_sample(const Scd41Sample& sample,
                                         float min_ppm,
                                         float max_ppm,
                                         uint32_t stale_after_ms) {
  if (!sample.communication_ok) return SensorSampleStatus::kMissing;
  if (!sample.data_ready) {
    return sample.age_ms >= stale_after_ms ? SensorSampleStatus::kStale : SensorSampleStatus::kNotReady;
  }
  if (!sample.crc_valid) return SensorSampleStatus::kCrcError;
  if (!std::isfinite(sample.co2_ppm) || sample.co2_ppm < min_ppm || sample.co2_ppm > max_ppm) {
    return SensorSampleStatus::kRange;
  }
  if (sample.age_ms >= stale_after_ms) return SensorSampleStatus::kStale;
  return SensorSampleStatus::kValid;
}

}  // namespace ventilation

File: 06 - LOGICIEL EMBARQUE/simulateur/src/hardware-contracts.js
const clamp = (value, min, max) => Math.min(max, Math.max(min, value));

export function dacWordFromPercent(percent) {
  const bounded = clamp(Number(percent), 0, 100);
  const millivolts = bounded * 100;
  const code12 = Math.trunc((millivolts / 10000) * 4095);
  return code12 << 4;
}

export function dacFrameFromPercent(percent, channel = 0) {
  const word = dacWordFromPercent(percent);
  const low = word & 0xff;
  const high = (word >> 8) & 0xff;
  if (channel === 0) return { register: 0x02, bytes: [low, high] };
  if (channel === 1) return { register: 0x04, bytes: [low, high] };
  if (channel === 2) return { register: 0x02, bytes: [low, high, low, high] };
  return { register: 0x00, bytes: [] };
}

export const SensorSampleStatus = Object.freeze({
  VALID: "VALID",
  NOT_READY: "NOT_READY",
  MISSING: "MISSING",
  CRC_ERROR: "CRC_ERROR",
  STALE: "STALE",
  RANGE: "RANGE",
});

export function classifyScd41Sample(sample, limits = {}) {
  const config = { minPpm: 350, maxPpm: 5000, staleAfterMs: 30000, ...limits };
  if (!sample.communicationOk) return SensorSampleStatus.MISSING;
  if (!sample.dataReady) return sample.ageMs >= config.staleAfterMs ? SensorSampleStatus.STALE : SensorSampleStatus.NOT_READY;
  if (!sample.crcValid) return SensorSampleStatus.CRC_ERROR;
  if (!Number.isFinite(sample.co2Ppm) || sample.co2Ppm < config.minPpm || sample.co2Ppm > config.maxPpm) return SensorSampleStatus.RANGE;
  if (sample.ageMs >= config.staleAfterMs) return SensorSampleStatus.STALE;
  return SensorSampleStatus.VALID;
}
